File: hard/int_palindrome.rb
def palindrome?(a)
  a == a.to_s.reverse.to_i
end

File.open(ARGV[0]).each_line do |line|
  l, r = line.split.map(&:to_i)
  n = 0
  (l..r).each do |i|
    prev = -1
    (i..r).each do |j|
      if prev >= 0
        p = prev
        p += 1 if palindrome?(j)
      else
        p = 0
        (i..j).each do |k|
          p += 1 if palindrome?(k)
        end
      end
      n += 1 if p.even?
      prev = p
    end
  end
  puts n
end

File: hard/int_palindrome.pl
use strict;
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    $line =~ /(\d+) (\d+)/;
    my ( $n, $l, $r ) = ( 0, $1, $2 );
    for ( my $i = $l ; $i <= $r ; $i++ ) {
        my $prev = -1;
        for ( my $j = $i ; $j <= $r ; $j++ ) {
            my $p;
            if ( $prev > 0 ) {
                $p = $prev;
                $p++ if ( $j eq reverse($j) );
            }
            else {
                $p = 0;
                for ( my $k = $i ; $k <= $j ; $k++ ) {
                    $p++ if ( $k eq reverse($k) );
                }
            }
            $n++ if ( ( $p & 1 ) == 0 );
            $prev = $p;
        }
    }
    print "$n\n";
}
close(INFILE);
